mon: add support public_bind_addr option

bassam opened the pull request:
To support running in dynamic enviornments (like Kubernetes), its desirable for a mon to have two ip addresses. The first it can bind to locally and the other is what goes in the monmap for others to reach the mon.

Added a new config option "public_bind_addr" which if set will be used by the mon to listen locally. The existing "public_addr" is unchanged and can be different from "public_bind_addr".

Also added a new function on Messenger to set_addr which is called by ceph-mon to set
the advertised address after doing the bind.

What's left:

- [x] add new unit tests
- [x] end to end testing with Rook
- [x] add docs

@liewegas would good to get early feedback on the approach here.

bassam commented on src/ceph_mon.cc:
```diff
@@ -727,12 +779,6 @@ int main(int argc, const char **argv)
     prefork.exit(1);
   }
 
-  dout(0) << "starting " << g_conf->name << " rank " << rank
```
I removed this since it was duplicative with the log message on line 757 above. not sure I understand why we need two log messages.

bassam commented on src/msg/xio/XioMessenger.h:
```diff
@@ -100,6 +100,8 @@ class XioMessenger : public SimplePolicyMessenger, XioInit
   virtual void set_addr_unknowns(const entity_addr_t &addr) override
     { } /* XXX applicable? */
+  virtual void set_addr(const entity_addr_t &addr) override
+    { } /* XXX applicable? */
```
I'm unable to build XioMessenger. The code looks like its gone stale.

bassam commented:
retest this please

liewegas commented:
This looks totally reasonable to me.  Let me know how testing goes!  Would be great to include this in luminous.

bassam commented:
@liewegas I reverted back to using the original "bind_addr" due to a number of setbacks (tests make lots of assumptions about mon hosts, and a number of corner cases that came up while testing).  The end result looks a lot better and much less churn. Please take a look.

Also added new unit tests that simulate the behavior of "routing" traffic from the public_addr to the public_bind_addr. it did require adding `socat` as a build dependency.

bassam commented on qa/workunits/ceph-helpers.sh:
```diff
@@ -1763,6 +1773,46 @@ if test "$1" = TESTS ; then
     run_tests "$@"
 fi
 
+# NOTE:
```
this moved here from src/test/mon/misc.sh

bassam commented on src/common/config_opts.h:
```diff
@@ -16,6 +16,7 @@
 OPTION(host, OPT_STR, "") // "" means that ceph will use short hostname
 OPTION(fsid, OPT_UUID, uuid_d())
 OPTION(public_addr, OPT_ADDR, entity_addr_t())
+OPTION(public_bind_addr, OPT_ADDR, entity_addr_t())
```
we could make this "mon_public_bind_addr" to be more specific. the same behavior could be carried over to OSD, MDS etc. if we wanted to so I left a bit more general. Note that we could also add "cluster_bind_addr" etc. if we wanted to too.

bassam commented on src/msg/async/AsyncConnection.cc:
```diff
@@ -977,9 +977,8 @@ ssize_t AsyncConnection::_process_connection()
-            ldout(async_msgr->cct, 0) << __func__ << " connect claims to be "
-                                << paddr << " not " << peer_addr << " - wrong node!" << dendl;
-            goto fail;
+            ldout(async_msgr->cct, 20) << __func__ << " connect claims to be "
```
I think its a little noisy too -- happy to remove it. Maybe @liewegas can weigh in here.

bassam commented:
@liewegas this is ready for review and testing. we ran this in Rook/Kuberentes scenarios and its working well.

liewegas commented:
retest this please

liewegas commented:
it looks like a submodule was inadvertantly updated?

bassam commented:
@liewegas oops. fixed.

bassam commented:
@liewegas just to confirm, @travisn is recommending we remove the warning completely since its now could potentially be normal operation and is likely to fill up the mon logs. what do you think?

liewegas commented:
Hmm, actually, let's leave it but make the level 10 and change the message 
to be something like "note peer addr does not match socket addr 
(peer using public_bind_addr?)"

liewegas commented:
http://pulpito.ceph.com/sage-2017-07-15_01:35:06-rados-wip-sage-testing-distro-basic-smithi/

smithfarm commented on ceph.spec.in:
```diff
@@ -131,6 +131,7 @@ BuildRequires:	python-pecan
 BuildRequires:	python-requests
 BuildRequires:	python-virtualenv
 BuildRequires:	python-werkzeug
+BuildRequires:	socat
```
@bassam This (socat) is only needed for the unit test, correct?

tchaikov replied:
@smithfarm i am not Bassam. but i agree that socat is only used for tests. is there anything we can do different? if it's only for test?

smithfarm replied:
@tchaikov: yes, will handle in #15940

bassam replied:
@smithfarm yes just for make check. Feel free to make it part of a conditional. Not it's also could be a conditional for debs and FreeBSD too.

smithfarm replied:
Conditionalizing build dependencies is far beyond my poor-to-nonexistent Debian and FreeBSD packaging skills.

tchaikov replied:
maybe we should just move the build-depends for readying "make check" into `install-deps.sh`.

smithfarm replied:
@tchaikov Theoretically, "make check" can be run by `rpmbuild`, though?

smithfarm replied:
see https://github.com/ceph/ceph/blob/master/ceph.spec.in#L860-L866

tchaikov replied:
@smithfarm sorry for the confusion, i was talking about the debian packaging.

smithfarm replied:
@tchaikov I see what you mean, now, and I think it's a good idea. One possible danger, though - people might forget that new unit test dependencies need to be added to ceph.spec.in for RPM, but to install-deps.sh for DEB. It's somewhat non-intuitive.
